README.md
# Weekly Date Picker

A weekday picker where you can scroll between weeks.

<img src="https://raw.githubusercontent.com/Magnuti/Weekly-Date-Picker/main/assets/white_with_week.gif">

## Custom styling

### You can remove the week text

<img src="https://raw.githubusercontent.com/Magnuti/Weekly-Date-Picker/main/assets/white_without_week.jpg" width="590">

### Add custom colors

<img src="https://raw.githubusercontent.com/Magnuti/Weekly-Date-Picker/main/assets/dark_without_week.jpg" width="590">

### You can also make only the weekdays selectable

<img src="https://raw.githubusercontent.com/Magnuti/Weekly-Date-Picker/main/assets/dark_five_days.jpg" width="590">

## Installing

To use this package, add `weekly_date_picker` as a dependency in your `pubspec.yaml` file.

## Usage

```dart
WeeklyDatePicker(
  selectedDay: _selectedDay, // DateTime
  changeDay: (value) => setState(() {
    _selectedDay = value;
  }),
),
```

## How to use custom styling

You can use custom colors and labels by the optional parameters:

- `weekdayText` Specifies the weekday text: default is 'Week'
- `weekdays` Specifies the weekday strings ['Mon', 'Tue'...]
- `backgroundColor` Background color
- `selectedDigitBackgroundColor` Color of the selected day circle
- `selectedDigitBorderColor` Color of the border of the selected day circle
- `selectedDigitColor` Color of the selected digit text
- `digitsColor` Color of the unselected digits text
- `weekdayTextColor` Is the color of the weekdays 'Mon', 'Tue'...
- `enableWeeknumberText` Set to false to hide the weeknumber textfield to the left of the slider
- `weeknumberColor` Color of the weekday container
- `weeknumberTextColor` Color of the weekday text
- `daysInWeek` Specifies the number of weekdays to render, default is 7, so Monday to Sunday


CHANGELOG.md
## 1.4.2

Small README update.

## 1.4.1

No functional changes. Fixed `CHANGELOG.md` to format correctly.

## 1.4.0

Added possibility to change the border color of the selected digit.

- Renamed `selectedBackgroundColor` to `selectedDigitBackgroundColor`

## 1.3.2

Fixed a bug with daylight savings causing the next day to not be calculated properly.

## 1.3.1

Upgraded Flutter version to 3.

## 1.3.0

Made it possible to swipe backwards in time from today's date.

## 1.2.0

- Replaced custom weeknumber calculation (which didn't work with leap years) with the `week_of_year` package.
- Fixed proper dispose of controller
- Added tests

## 1.1.3

Fixed wrong naming due to previous package name

## 1.1.2

Updated README.

## 1.1.1

Changes package name.

## 1.1.0

- Migrated to Flutter 2
- Added support for disabling the weekday text
- Adjusted borders, margins and colors
- Improved parameter names

## 1.0.2

Fixed GIFs for pub.dev README

## 1.0.1

Added documentation

## 1.0.0

Initial version


test/datetime_apis_test.dart
import 'package:flutter_test/flutter_test.dart';
import "package:weekly_date_picker/datetime_apis.dart";

void main() {
  group("DateTime extension should compare dates only ", () {
    test("Dates should be equal", () {
      DateTime d1 = new DateTime(2020, 2, 2);
      DateTime d2 = new DateTime(2020, 2, 2);

      expect(d1.isSameDateAs(d2), true);
    });

    test("Dates should be equal", () {
      DateTime d1 = new DateTime(2020, 2, 2, 17, 30);
      DateTime d2 = new DateTime(2020, 2, 2, 10, 20);

      expect(d1.isSameDateAs(d2), true);
    });

    test("Dates should be unequal", () {
      DateTime d1 = new DateTime(2020, 2, 3);
      DateTime d2 = new DateTime(2020, 2, 2);

      expect(d1.isSameDateAs(d2), false);
    });

    test("Dates should be unequal", () {
      DateTime d1 = new DateTime(2020, 5, 2);
      DateTime d2 = new DateTime(2020, 1, 2);

      expect(d1.isSameDateAs(d2), false);
    });

    test("Dates should be unequal", () {
      DateTime d1 = new DateTime(2019, 1, 1, 17, 30);
      DateTime d2 = new DateTime(2020, 1, 1, 10, 20);

      expect(d1.isSameDateAs(d2), false);
    });
  });

  group("DateTime addDate extension should work ", () {
    test("Add dates should work", () {
      DateTime date = new DateTime(2023, 1, 1);
      expect(date.addDays(1).isSameDateAs(DateTime(2023, 1, 2)), true);
      expect(date.addDays(30).isSameDateAs(DateTime(2023, 1, 31)), true);
      expect(date.addDays(31).isSameDateAs(DateTime(2023, 2, 1)), true);
    });

    test("Add dates should work with daylight savings", () {
      DateTime date = new DateTime(2023, 10, 29);
      expect(date.addDays(1).isSameDateAs(DateTime(2023, 10, 30)), true);

      // This one should fail since that date is daylight savings in Europe
      expect(date.add(Duration(days: 1)).isSameDateAs(DateTime(2023, 10, 30)),
          false);
    });

    test("Add dates should works over new year", () {
      DateTime date = new DateTime(2020, 12, 30);
      expect(date.addDays(2).isSameDateAs(DateTime(2021, 1, 1)), true);
    });
  });
}


test/weekday_scroller_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:weekday_scroller/weekday_scroller.dart';

void main() {
  test('adds one to input values', () {
    expect(1, 1);
  });
}


test/weekly_date_picker_test.dart
import 'package:clock/clock.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:weekly_date_picker/weekly_date_picker.dart';

void main() {
  final DateTime defaultTime = DateTime(2021, 09, 28, 14, 0);
  final List<int> defaultDates = [27, 28, 29, 30, 1, 2, 3];
  final List<String> defaultWeekdays = [
    'Mon',
    'Tue',
    'Wed',
    'Thu',
    'Fri',
    'Sat',
    'Sun'
  ];

  testWidgets("WeeklyDatePicker works with MaterialApp",
      (WidgetTester tester) async {
    withClock(Clock.fixed(defaultTime), () async {
      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
                selectedDay: clock.now(), changeDay: (value) => {})),
      );
      final weekTextFinder = find.text("Week 39");
      expect(weekTextFinder, findsOneWidget);

      for (String weekday in defaultWeekdays) {
        final weekdayFinder = find.text(weekday);
        expect(weekdayFinder, findsOneWidget);
      }

      for (int date in defaultDates) {
        final dateFinder = find.text(date.toString());
        expect(dateFinder, findsOneWidget);
      }
    });
  });

  testWidgets("WeeklyDatePicker works with CupertinoApp",
      (WidgetTester tester) async {
    withClock(Clock.fixed(defaultTime), () async {
      await tester.pumpWidget(
        CupertinoApp(
            home: WeeklyDatePicker(
                selectedDay: clock.now(), changeDay: (value) => {})),
      );

      final weekTextFinder = find.text("Week 39");
      expect(weekTextFinder, findsOneWidget);

      for (String weekday in defaultWeekdays) {
        final weekdayFinder = find.text(weekday);
        expect(weekdayFinder, findsOneWidget);
      }

      for (int date in defaultDates) {
        final dateFinder = find.text(date.toString());
        expect(dateFinder, findsOneWidget);
      }
    });
  });

  testWidgets(
      "WeeklyDatePicker shows all custom weekdays with custom week length",
      (WidgetTester tester) async {
    withClock(Clock.fixed(defaultTime), () async {
      final List<String> weekdays = ['Mo', 'Tu', 'We', 'Th', 'Fr'];
      final List<int> dates = [27, 28, 29, 30, 1];

      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
          selectedDay: clock.now(),
          changeDay: (value) => {},
          weekdays: weekdays,
          daysInWeek: 5,
        )),
      );

      for (String weekday in weekdays) {
        final weekdayFinder = find.text(weekday);
        expect(weekdayFinder, findsOneWidget);
      }

      for (int date in dates) {
        final dateFinder = find.text(date.toString());
        expect(dateFinder, findsOneWidget);
      }
    });
  });

  testWidgets("WeeklyDatePicker correctly does not display the week text",
      (WidgetTester tester) async {
    withClock(Clock.fixed(defaultTime), () async {
      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
          selectedDay: clock.now(),
          changeDay: (value) => {},
          enableWeeknumberText: false,
        )),
      );

      final weekTextFinder = find.text("Week 39");
      expect(weekTextFinder, findsNothing);
    });
  });

  testWidgets("WeeklyDatePicker allows custom week text",
      (WidgetTester tester) async {
    withClock(Clock.fixed(defaultTime), () async {
      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
          selectedDay: clock.now(),
          changeDay: (value) => {},
          weekdayText: "Uke",
        )),
      );

      final weekTextFinder = find.text("Week 39");
      expect(weekTextFinder, findsNothing);
      final ukeTextFinder = find.text("Uke 39");
      expect(ukeTextFinder, findsOneWidget);
    });
  });

  testWidgets("WeeklyDatePicker works with swipe", (WidgetTester tester) async {
    withClock(Clock.fixed(defaultTime), () async {
      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
                selectedDay: clock.now(), changeDay: (value) => {})),
      );

      final pageViewFinder = find.byType(PageView);
      final nextWeekDateFinder = find.text("4");

      await tester.dragUntilVisible(
        nextWeekDateFinder,
        pageViewFinder,
        Offset(-250, 0),
      );

      expect(nextWeekDateFinder, findsOneWidget);
    });
  });

  testWidgets("WeeklyDatePicker works with leap years",
      (WidgetTester tester) async {
    final DateTime leapYearDateTime = DateTime(2020, 02, 27);

    withClock(Clock.fixed(leapYearDateTime), () async {
      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
                selectedDay: leapYearDateTime, changeDay: (value) => {})),
      );

      final List<int> dates = [24, 25, 26, 27, 28, 29, 1];

      for (int date in dates) {
        final dateFinder = find.text(date.toString());
        expect(dateFinder, findsOneWidget);
      }

      expect(find.text("Week 9"), findsOneWidget);
    });
  });

  testWidgets("WeeklyDatePicker works over new year",
      (WidgetTester tester) async {
    final DateTime leapYearDateTime = DateTime(2020, 12, 30);

    withClock(Clock.fixed(leapYearDateTime), () async {
      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
                selectedDay: leapYearDateTime, changeDay: (value) => {})),
      );

      final List<int> dates = [28, 29, 30, 31, 1, 2, 3];

      for (int date in dates) {
        final dateFinder = find.text(date.toString());
        expect(dateFinder, findsOneWidget);
      }

      expect(find.text("Week 53"), findsOneWidget);

      final pageViewFinder = find.byType(PageView);
      final nextWeekDateFinder = find.text("10");

      await tester.dragUntilVisible(
        nextWeekDateFinder,
        pageViewFinder,
        Offset(-250, 0),
      );

      expect(find.text("Week 54"), findsNothing);
      expect(find.text("Week 1"), findsOneWidget);

      final List<int> datesAfterScroll = [4, 5, 6, 7, 8, 9, 10];

      for (int date in datesAfterScroll) {
        final dateFinder = find.text(date.toString());
        expect(dateFinder, findsOneWidget);
      }
    });
  });

  testWidgets("WeeklyDatePicker works backwards in time",
      (WidgetTester tester) async {
    withClock(Clock.fixed(defaultTime), () async {
      await tester.pumpWidget(
        MaterialApp(
            home: WeeklyDatePicker(
                selectedDay: clock.now(), changeDay: (value) => {})),
      );

      expect(find.text("Week 39"), findsOneWidget);

      final pageViewFinder = find.byType(PageView);
      final nextWeekDateFinder = find.text("21");

      await tester.dragUntilVisible(
        nextWeekDateFinder,
        pageViewFinder,
        Offset(250, 0),
      );

      expect(nextWeekDateFinder, findsOneWidget);

      final List<int> previousWeekDates = [20, 21, 22, 23, 24, 25, 26];
      for (int date in previousWeekDates) {
        final dateFinder = find.text(date.toString());
        expect(dateFinder, findsOneWidget);
      }

      expect(find.text("Week 38"), findsOneWidget);
      expect(find.text("Week 39"), findsNothing);
    });
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:weekly_date_picker/weekly_date_picker.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Weekly Date Picker Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: const MyHomePage(title: 'Weekly Date Picker Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  MyHomePageState createState() => MyHomePageState();
}

class MyHomePageState extends State<MyHomePage> {
  DateTime _selectedDay = DateTime.now();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          WeeklyDatePicker(
            selectedDay: _selectedDay,
            changeDay: (value) => setState(() {
              _selectedDay = value;
            }),
            enableWeeknumberText: false,
            weeknumberColor: const Color(0xFF57AF87),
            weeknumberTextColor: Colors.white,
            backgroundColor: const Color(0xFF1A1A1A),
            weekdayTextColor: const Color(0xFF8A8A8A),
            digitsColor: Colors.white,
            selectedDigitBackgroundColor: const Color(0xFF57AF87),
            weekdays: const ["Mo", "Tu", "We", "Th", "Fr"],
            daysInWeek: 5,
          ),
        ],
      ),
    );
  }
}


lib/weekday_scroller.dart
library day_scroller;

import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

class WeekdayScroller extends StatefulWidget {
  WeekdayScroller({
    Key? key,
    required this.selectedDay,
    required this.changeDay,
    this.weekdayText = 'Week',
    this.weekdays = const ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'],
    this.backgroundColor = const Color(0xFFFAFAFA),
    this.selectedBackgroundColor = const Color(0xFF2A2859),
    this.selectedDigitColor = const Color(0xFFFFFFFF),
    this.digitsColor = const Color(0xFF000000),
    this.weekdayTextColor = const Color(0xFF303030),
    this.enableWeeknumberText = true,
    this.weeknumberColor = const Color(0xFFB2F5FE),
    this.weeknumberTextColor = const Color(0xFF000000),
    this.daysInWeek = 7,
  })  : assert(weekdays.length == daysInWeek,
            "weekdays must be of length $daysInWeek"),
        super(key: key);

  /// The current selected day
  final DateTime selectedDay;

  /// Callback function with the new selected date
  final Function(DateTime) changeDay;

  /// Specifies the weekday text: default is 'Week'
  final String weekdayText;

  /// Specifies the weekday strings ['Mon', 'Tue'...]
  final List<String> weekdays;

  /// Background color
  final Color backgroundColor;

  /// Color of the selected digits text
  final Color selectedBackgroundColor;

  /// Color of the unselected digits text
  final Color selectedDigitColor;

  /// Color of the unselected digits text
  final Color digitsColor;

  /// Is the color of the weekdays 'Mon', 'Tue'...
  final Color weekdayTextColor;

  /// Set to false to hide the weeknumber textfield to the left of the slider
  final bool enableWeeknumberText;

  /// Color of the weekday container
  final Color weeknumberColor;

  /// Color of the weekday text
  final Color weeknumberTextColor;

  /// Specifies the number of weekdays to render, default is 7, so Monday to Sunday
  final int daysInWeek;

  @override
  _WeekdayScrollerState createState() => _WeekdayScrollerState();
}

class _WeekdayScrollerState extends State<WeekdayScroller> {
  final controller = PageController(initialPage: 0);
  final DateTime now = DateTime.now();

  late int _weeknumberNow;
  late int _weeknumberInSwipe;

  @override
  void initState() {
    super.initState();
    // Weekday calculation from https://en.wikipedia.org/wiki/ISO_week_date#Calculation
    int dayOfYear =
        int.parse(DateFormat('D').format(now)); // day count from. 1. January
    _weeknumberNow = ((dayOfYear - now.weekday + 10) / 7).floor();
    _weeknumberInSwipe = _weeknumberNow;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 64,
      color: widget.backgroundColor,
      child: Row(
        children: <Widget>[
          widget.enableWeeknumberText
              ? Container(
                  padding: EdgeInsets.all(8.0),
                  color: widget.weeknumberColor,
                  child: Text(
                    '${widget.weekdayText} $_weeknumberInSwipe',
                    style: TextStyle(color: widget.weeknumberTextColor),
                  ),
                )
              : Container(),
          Expanded(
            child: PageView.builder(
              controller: controller,
              onPageChanged: (int index) {
                setState(() {
                  _weeknumberInSwipe = _weeknumberNow + index;
                });
              },
              scrollDirection: Axis.horizontal,
              itemBuilder: (_, weekOffset) => Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: _weekdays(weekOffset),
              ),
            ),
          ),
        ],
      ),
    );
  }

  // Builds a 5 day list of DateButtons
  List<Widget> _weekdays(int weekIndex) {
    List<Widget> weekdays = [];

    for (int i = 0; i < widget.daysInWeek; i++) {
      int offset = i + 1 - now.weekday;
      DateTime dateTime = now.add(Duration(days: weekIndex * 7 + offset));
      weekdays.add(_dateButton(dateTime));
    }
    return weekdays;
  }

  Widget _dateButton(DateTime dateTime) {
    final String weekday = widget.weekdays[dateTime.weekday - 1];
    final bool isSelected = _isSameDate(dateTime, widget.selectedDay);

    return Expanded(
      child: GestureDetector(
        onTap: () => widget.changeDay(dateTime),
        child: Container(
          // Bugfix, the transparent container makes the GestureDetector fill the Expanded
          color: Colors.transparent,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Padding(
                padding: EdgeInsets.only(bottom: 4.0),
                child: Text(
                  '$weekday',
                  style:
                      TextStyle(fontSize: 12.0, color: widget.weekdayTextColor),
                ),
              ),
              Container(
                padding: const EdgeInsets.all(1.0),
                decoration: BoxDecoration(
                    // Border around today's date
                    color: _isSameDate(dateTime, now)
                        ? widget.selectedBackgroundColor
                        : Colors.transparent,
                    shape: BoxShape.circle),
                child: CircleAvatar(
                  backgroundColor: isSelected
                      ? widget.selectedBackgroundColor
                      : widget.backgroundColor,
                  radius: 14.0,
                  child: Text(
                    '${dateTime.day}',
                    style: TextStyle(
                        fontSize: 16.0,
                        color: isSelected
                            ? widget.selectedDigitColor
                            : widget.digitsColor),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  bool _isSameDate(DateTime d1, DateTime d2) {
    return d1.day == d2.day && d1.month == d2.month && d1.year == d2.year;
  }
}


lib/weekly_date_picker.dart
library weekly_date_picker;

import 'package:flutter/material.dart';
import 'package:week_of_year/week_of_year.dart';
import "package:weekly_date_picker/datetime_apis.dart";

class WeeklyDatePicker extends StatefulWidget {
  WeeklyDatePicker({
    Key? key,
    required this.selectedDay,
    required this.changeDay,
    this.weekdayText = 'Week',
    this.weekdays = const ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'],
    this.backgroundColor = const Color(0xFFFAFAFA),
    this.selectedDigitBackgroundColor = const Color(0xFF2A2859),
    this.selectedDigitBorderColor =
        const Color(0x00000000), // Transparent color
    this.selectedDigitColor = const Color(0xFFFFFFFF),
    this.digitsColor = const Color(0xFF000000),
    this.weekdayTextColor = const Color(0xFF303030),
    this.enableWeeknumberText = true,
    this.weeknumberColor = const Color(0xFFB2F5FE),
    this.weeknumberTextColor = const Color(0xFF000000),
    this.daysInWeek = 7,
  })  : assert(weekdays.length == daysInWeek,
            "weekdays must be of length $daysInWeek"),
        super(key: key);

  /// The current selected day
  final DateTime selectedDay;

  /// Callback function with the new selected date
  final Function(DateTime) changeDay;

  /// Specifies the weekday text: default is 'Week'
  final String weekdayText;

  /// Specifies the weekday strings ['Mon', 'Tue'...]
  final List<String> weekdays;

  /// Background color
  final Color backgroundColor;

  /// Color of the selected day circle
  final Color selectedDigitBackgroundColor;

  /// Color of the border of the selected day circle
  final Color selectedDigitBorderColor;

  /// Color of the selected digit text
  final Color selectedDigitColor;

  /// Color of the unselected digits text
  final Color digitsColor;

  /// Is the color of the weekdays 'Mon', 'Tue'...
  final Color weekdayTextColor;

  /// Set to false to hide the weeknumber textfield to the left of the slider
  final bool enableWeeknumberText;

  /// Color of the weekday container
  final Color weeknumberColor;

  /// Color of the weekday text
  final Color weeknumberTextColor;

  /// Specifies the number of weekdays to render, default is 7, so Monday to Sunday
  final int daysInWeek;

  @override
  _WeeklyDatePickerState createState() => _WeeklyDatePickerState();
}

class _WeeklyDatePickerState extends State<WeeklyDatePicker> {
  final DateTime _todaysDateTime = DateTime.now();

  // About 100 years back in time should be sufficient for most users, 52 weeks * 100
  final int _weekIndexOffset = 5200;

  late final PageController _controller;
  late final DateTime _initialSelectedDay;
  late int _weeknumberInSwipe;

  @override
  void initState() {
    super.initState();
    _controller = PageController(initialPage: _weekIndexOffset);
    _initialSelectedDay = widget.selectedDay;
    _weeknumberInSwipe = widget.selectedDay.weekOfYear;
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 64,
      color: widget.backgroundColor,
      child: Row(
        children: <Widget>[
          widget.enableWeeknumberText
              ? Container(
                  padding: EdgeInsets.all(8.0),
                  color: widget.weeknumberColor,
                  child: Text(
                    '${widget.weekdayText} $_weeknumberInSwipe',
                    style: TextStyle(color: widget.weeknumberTextColor),
                  ),
                )
              : Container(),
          Expanded(
            child: PageView.builder(
              controller: _controller,
              onPageChanged: (int index) {
                setState(() {
                  _weeknumberInSwipe = _initialSelectedDay
                      .addDays(7 * (index - _weekIndexOffset))
                      .weekOfYear;
                });
              },
              scrollDirection: Axis.horizontal,
              itemBuilder: (_, weekIndex) => Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: _weekdays(weekIndex - _weekIndexOffset),
              ),
            ),
          ),
        ],
      ),
    );
  }

  // Builds a 5 day list of DateButtons
  List<Widget> _weekdays(int weekIndex) {
    List<Widget> weekdays = [];

    for (int i = 0; i < widget.daysInWeek; i++) {
      final int offset = i + 1 - _initialSelectedDay.weekday;
      final int daysToAdd = weekIndex * 7 + offset;
      final DateTime dateTime = _initialSelectedDay.addDays(daysToAdd);
      weekdays.add(_dateButton(dateTime));
    }
    return weekdays;
  }

  Widget _dateButton(DateTime dateTime) {
    final String weekday = widget.weekdays[dateTime.weekday - 1];
    final bool isSelected = dateTime.isSameDateAs(widget.selectedDay);
    final bool isTodaysDate = dateTime.isSameDateAs(_todaysDateTime);

    return Expanded(
      child: GestureDetector(
        onTap: () => widget.changeDay(dateTime),
        child: Container(
          // Bugfix, the transparent container makes the GestureDetector fill the Expanded
          color: Colors.transparent,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Padding(
                padding: EdgeInsets.only(bottom: 4.0),
                child: Text(
                  '$weekday',
                  style:
                      TextStyle(fontSize: 12.0, color: widget.weekdayTextColor),
                ),
              ),
              Container(
                padding: const EdgeInsets.all(1.0),
                decoration: BoxDecoration(
                    // Border around today's date
                    color: isTodaysDate
                        ? widget.selectedDigitBorderColor
                        : Colors.transparent,
                    shape: BoxShape.circle),
                child: CircleAvatar(
                  backgroundColor: isSelected
                      ? widget.selectedDigitBackgroundColor
                      : widget.backgroundColor,
                  radius: 14.0,
                  child: Text(
                    '${dateTime.day}',
                    style: TextStyle(
                        fontSize: 16.0,
                        color: isSelected
                            ? widget.selectedDigitColor
                            : widget.digitsColor),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
